File: sdks/typescript/src/websocket.ts
/**
 * UPP SDK WebSocket Client
 *
 * Real-time subscriptions to market prices and orderbook updates.
 * Includes automatic reconnection with exponential backoff.
 */

import * as types from './types';

/**
 * Configuration for the WebSocket client.
 */
export interface WebSocketClientConfig {
  /** WebSocket URL (e.g., "ws://localhost:8080/upp/v1/ws") */
  url: string;
  /** Reconnection options */
  reconnect?: {
    enabled: boolean;
    maxAttempts: number;
    initialDelayMs: number;
    maxDelayMs: number;
    backoffMultiplier: number;
  };
  /** Heartbeat interval in milliseconds (default: 30000) */
  heartbeatInterval?: number;
  /** Custom WebSocket implementation */
  WebSocket?: typeof globalThis.WebSocket;
}

/**
 * Callbacks for WebSocket events.
 */
export interface WebSocketCallbacks {
  /** Called when connected */
  onConnect?: () => void;
  /** Called when disconnected */
  onDisconnect?: () => void;
  /** Called on price updates */
  onPrice?: (update: types.PriceUpdate) => void;
  /** Called on orderbook updates */
  onOrderbook?: (update: types.OrderbookUpdate) => void;
  /** Called on error */
  onError?: (error: Error) => void;
}

/**
 * WebSocket client for real-time market data.
 *
 * @example
 * ```typescript
 * const ws = new UppWebSocket({
 *   url: 'ws://localhost:8080/upp/v1/ws'
 * });
 *
 * ws.on({
 *   onPrice: (update) => console.log('Price:', update),
 *   onOrderbook: (update) => console.log('Orderbook:', update),
 * });
 *
 * await ws.connect();
 * await ws.subscribePrices(['upp:kalshi:MELON-240301']);
 * ```
 */
export class UppWebSocket {
  private url: string;
  private socket?: WebSocket;
  private reconnectConfig: Required<NonNullable<WebSocketClientConfig['reconnect']>>;
  private heartbeatInterval: number;
  private WebSocketImpl: typeof globalThis.WebSocket;
  private callbacks: WebSocketCallbacks = {};
  private reconnectAttempts = 0;
  private heartbeatHandle?: NodeJS.Timeout;
  private reconnectHandle?: NodeJS.Timeout;
  private connected = false;
  private subscriptions = new Map<string, Set<string>>();
  private requestId = 0;

  constructor(config: WebSocketClientConfig) {
    this.url = config.url;
    this.heartbeatInterval = config.heartbeatInterval ?? 30000;
    this.WebSocketImpl = config.WebSocket ?? globalThis.WebSocket;

    this.reconnectConfig = {
      enabled: config.reconnect?.enabled ?? true,
      maxAttempts: config.reconnect?.maxAttempts ?? 10,
      initialDelayMs: config.reconnect?.initialDelayMs ?? 1000,
      maxDelayMs: config.reconnect?.maxDelayMs ?? 30000,
      backoffMultiplier: config.reconnect?.backoffMultiplier ?? 2,
    };
  }

  /**
   * Register event callbacks.
   *
   * @param callbacks - Callback handlers
   */
  on(callbacks: WebSocketCallbacks): void {
    this.callbacks = { ...this.callbacks, ...callbacks };
  }

  /**
   * Connect to the WebSocket server.
   *
   * @returns Promise that resolves when connected
   */
  async connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        this.socket = new this.WebSocketImpl(this.url);

        this.socket.onopen = () => {
          this.connected = true;
          this.reconnectAttempts = 0;
          this.startHeartbeat();
          this.callbacks.onConnect?.();
          resolve();
        };

        this.socket.onmessage = (event) => {
          this.handleMessage(event.data);
        };

        this.socket.onerror = (error) => {
          const err = new Error(`WebSocket error: ${error}`);
          this.callbacks.onError?.(err);
          reject(err);
        };

        this.socket.onclose = () => {
          this.connected = false;
          this.stopHeartbeat();
          this.callbacks.onDisconnect?.();
          this.attemptReconnect();
        };
      } catch (error) {
        reject(error);
      }
    });
  }

  /**
   * Disconnect from the WebSocket server.
   */
  disconnect(): void {
    this.stopHeartbeat();
    this.clearReconnectHandle();
    if (this.socket) {
      this.socket.close();
      this.socket = undefined;
    }
    this.connected = false;
  }

  /**
   * Subscribe to price updates for specified markets.
   *
   * @param marketIds - Market IDs to subscribe to
   * @param intervalMs - Update interval in milliseconds (default: 1000)
   */
  async subscribePrices(marketIds: string[], intervalMs: number = 1000): Promise<void> {
    const id = ++this.requestId;
    const sub: types.PriceSubscription = {
      channel: 'prices',
      market_ids: marketIds,
      interval_ms: intervalMs,
    };

    const request: types.JsonRpcRequest = {
      jsonrpc: '2.0',
      id,
      method: 'subscribe_prices',
      params: sub,
    };

    return this.sendRequest(request, id);
  }

  /**
   * Subscribe to orderbook updates for specified markets.
   *
   * @param marketIds - Market IDs to subscribe to
   * @param depth - Orderbook depth (default: 10)
   * @param intervalMs - Update interval in milliseconds (default: 2000)
   */
  async subscribeOrderbook(
    marketIds: string[],
    depth: number = 10,
    intervalMs: number = 2000
  ): Promise<void> {
    const id = ++this.requestId;
    const sub: types.OrderbookSubscription = {
      channel: 'orderbook',
      market_ids: marketIds,
      depth,
      interval_ms: intervalMs,
    };

    const request: types.JsonRpcRequest = {
      jsonrpc: '2.0',
      id,
      method: 'subscribe_orderbook',
      params: sub,
    };

    return this.sendRequest(request, id);
  }

  /**
   * Unsubscribe from a channel.
   *
   * @param channel - Channel name ("prices" or "orderbook")
   * @param marketIds - Market IDs to unsubscribe from
   */
  async unsubscribe(channel: string, marketIds: string[]): Promise<void> {
    const id = ++this.requestId;
    const request: types.JsonRpcRequest = {
      jsonrpc: '2.0',
      id,
      method: 'unsubscribe',
      params: {
        channel,
        market_ids: marketIds,
      },
    };

    return this.sendRequest(request, id);
  }

  /**
   * Get a specific market (one-off request).
   *
   * @param marketId - Market ID
   * @returns Market data
   */
  async getMarket(marketId: string): Promise<types.Market> {
    const id = ++this.requestId;
    const request: types.JsonRpcRequest = {
      jsonrpc: '2.0',
      id,
      method: 'get_market',
      params: { market_id: marketId },
    };

    return this.sendRpcRequest(request, id);
  }

  /**
   * Check if WebSocket is connected.
   *
   * @returns Connection state
   */
  isConnected(): boolean {
    return this.connected && this.socket?.readyState === 1;
  }

  /**
   * Get current subscriptions.
   *
   * @returns Subscriptions map
   */
  getSubscriptions(): Map<string, Set<string>> {
    return new Map(this.subscriptions);
  }

  // ─── Private Methods ────────────────────────────────────

  /**
   * Send a JSON-RPC request and wait for response.
   *
   * @internal
   */
  private sendRequest(request: types.JsonRpcRequest, id: number | string): Promise<void> {
    return new Promise((resolve, reject) => {
      if (!this.socket || this.socket.readyState !== 1) {
        reject(new Error('WebSocket not connected'));
        return;
      }

      const timeoutHandle = setTimeout(() => {
        reject(new Error(`Request ${id} timed out`));
      }, 10000);

      const originalOnMessage = this.socket.onmessage;
      const handler = (event: MessageEvent) => {
        try {
          const response = JSON.parse(event.data) as types.JsonRpcResponse;
          if (response.id === id) {
            clearTimeout(timeoutHandle);
            if (response.error) {
              reject(new Error(response.error.message));
            } else {
              resolve();
            }
            // Restore original handler
            if (originalOnMessage) {
              this.socket!.onmessage = originalOnMessage;
            }
          }
        } catch (error) {
          // Not a JSON-RPC response, might be a fan-out message
        }
      };

      this.socket.onmessage = handler;
      this.socket.send(JSON.stringify(request));
    });
  }

  /**
   * Send a JSON-RPC request that expects a result.
   *
   * @internal
   */
  private sendRpcRequest<T>(request: types.JsonRpcRequest, id: number | string): Promise<T> {
    return new Promise((resolve, reject) => {
      if (!this.socket || this.socket.readyState !== 1) {
        reject(new Error('WebSocket not connected'));
        return;
      }

      const timeoutHandle = setTimeout(() => {
        reject(new Error(`Request ${id} timed out`));
      }, 10000);

      const originalOnMessage = this.socket.onmessage;
      const handler = (event: MessageEvent) => {
        try {
          const response = JSON.parse(event.data) as types.JsonRpcResponse;
          if (response.id === id) {
            clearTimeout(timeoutHandle);
            if (response.error) {
              reject(new Error(response.error.message));
            } else {
              resolve(response.result as T);
            }
            // Restore original handler
            if (originalOnMessage) {
              this.socket!.onmessage = originalOnMessage;
            }
          }
        } catch (error) {
          // Not a JSON-RPC response, might be a fan-out message
        }
      };

      this.socket.onmessage = handler;
      this.socket.send(JSON.stringify(request));
    });
  }

  /**
   * Handle incoming WebSocket messages.
   *
   * @internal
   */
  private handleMessage(data: string): void {
    try {
      const message = JSON.parse(data);

      // Check if it's a fan-out message or JSON-RPC response
      if ('channel' in message) {
        const fanOut = message as types.FanOutMessage;
        if (fanOut.channel === 'prices') {
          this.callbacks.onPrice?.(fanOut as types.PriceUpdate);
        } else if (fanOut.channel === 'orderbook') {
          this.callbacks.onOrderbook?.(fanOut as types.OrderbookUpdate);
        }
      }
    } catch (error) {
      this.callbacks.onError?.(new Error(`Failed to parse message: ${error}`));
    }
  }

  /**
   * Attempt to reconnect with exponential backoff.
   *
   * @internal
   */
  private attemptReconnect(): void {
    if (!this.reconnectConfig.enabled) return;
    if (this.reconnectAttempts >= this.reconnectConfig.maxAttempts) {
      this.callbacks.onError?.(
        new Error(`Failed to reconnect after ${this.reconnectAttempts} attempts`)
      );
      return;
    }

    const delayMs = Math.min(
      this.reconnectConfig.initialDelayMs *
        Math.pow(this.reconnectConfig.backoffMultiplier, this.reconnectAttempts),
      this.reconnectConfig.maxDelayMs
    );

    this.reconnectHandle = setTimeout(() => {
      this.reconnectAttempts++;
      this.connect().catch((error) => {
        this.callbacks.onError?.(error);
      });
    }, delayMs);
  }

  /**
   * Start heartbeat timer.
   *
   * @internal
   */
  private startHeartbeat(): void {
    this.heartbeatHandle = setInterval(() => {
      if (this.isConnected()) {
        const id = ++this.requestId;
        const request: types.JsonRpcRequest = {
          jsonrpc: '2.0',
          id,
          method: 'ping',
          params: {},
        };
        this.socket?.send(JSON.stringify(request));
      }
    }, this.heartbeatInterval);
  }

  /**
   * Stop heartbeat timer.
   *
   * @internal
   */
  private stopHeartbeat(): void {
    if (this.heartbeatHandle) {
      clearInterval(this.heartbeatHandle);
      this.heartbeatHandle = undefined;
    }
  }

  /**
   * Clear reconnect handle.
   *
   * @internal
   */
  private clearReconnectHandle(): void {
    if (this.reconnectHandle) {
      clearTimeout(this.reconnectHandle);
      this.reconnectHandle = undefined;
    }
  }
}

File: sdks/python/upp/websocket.py
"""
UPP SDK WebSocket Client

Real-time subscriptions to market prices and orderbook updates.
Includes automatic reconnection with exponential backoff.
"""

import asyncio
import json
from typing import Any, Callable, Dict, Optional

import websockets
from websockets.client import WebSocketClientProtocol

from .types import (
    JsonRpcRequest,
    JsonRpcResponse,
    MarketStatus,
    OrderbookSubscription,
    OrderbookUpdate,
    PriceSubscription,
    PriceUpdate,
)


class UppWebSocket:
    """
    WebSocket client for real-time market data.

    Example:
        >>> ws = UppWebSocket(url='ws://localhost:8080/upp/v1/ws')
        >>> ws.on_price(lambda u: print(f'Price: {u}'))
        >>> await ws.connect()
        >>> await ws.subscribe_prices(['upp:kalshi:MELON-240301'])
        >>> # ... later ...
        >>> ws.disconnect()
    """

    def __init__(
        self,
        url: str,
        reconnect: bool = True,
        max_reconnect_attempts: int = 10,
        initial_reconnect_delay: float = 1.0,
        max_reconnect_delay: float = 30.0,
        reconnect_backoff_multiplier: float = 2.0,
        heartbeat_interval: float = 30.0,
    ):
        """
        Initialize the WebSocket client.

        Args:
            url: WebSocket URL (e.g., 'ws://localhost:8080/upp/v1/ws')
            reconnect: Enable automatic reconnection
            max_reconnect_attempts: Maximum reconnection attempts
            initial_reconnect_delay: Initial reconnection delay in seconds
            max_reconnect_delay: Maximum reconnection delay in seconds
            reconnect_backoff_multiplier: Backoff multiplier for reconnection
            heartbeat_interval: Heartbeat interval in seconds
        """
        self.url = url
        self.reconnect_enabled = reconnect
        self.max_reconnect_attempts = max_reconnect_attempts
        self.initial_reconnect_delay = initial_reconnect_delay
        self.max_reconnect_delay = max_reconnect_delay
        self.reconnect_backoff_multiplier = reconnect_backoff_multiplier
        self.heartbeat_interval = heartbeat_interval

        self.socket: Optional[WebSocketClientProtocol] = None
        self.connected = False
        self.reconnect_attempts = 0

        self._on_connect: Optional[Callable[[], None]] = None
        self._on_disconnect: Optional[Callable[[], None]] = None
        self._on_price: Optional[Callable[[PriceUpdate], None]] = None
        self._on_orderbook: Optional[Callable[[OrderbookUpdate], None]] = None
        self._on_error: Optional[Callable[[Exception], None]] = None

        self._request_id = 0
        self._heartbeat_task: Optional[asyncio.Task[Any]] = None
        self._reconnect_task: Optional[asyncio.Task[Any]] = None
        self._receive_task: Optional[asyncio.Task[Any]] = None

    def on_connect(self, callback: Callable[[], None]) -> "UppWebSocket":
        """Register connect callback."""
        self._on_connect = callback
        return self

    def on_disconnect(self, callback: Callable[[], None]) -> "UppWebSocket":
        """Register disconnect callback."""
        self._on_disconnect = callback
        return self

    def on_price(
        self, callback: Callable[[PriceUpdate], None]
    ) -> "UppWebSocket":
        """Register price update callback."""
        self._on_price = callback
        return self

    def on_orderbook(
        self, callback: Callable[[OrderbookUpdate], None]
    ) -> "UppWebSocket":
        """Register orderbook update callback."""
        self._on_orderbook = callback
        return self

    def on_error(
        self, callback: Callable[[Exception], None]
    ) -> "UppWebSocket":
        """Register error callback."""
        self._on_error = callback
        return self

    async def connect(self) -> None:
        """Connect to the WebSocket server."""
        try:
            self.socket = await websockets.connect(self.url)
            self.connected = True
            self.reconnect_attempts = 0

            if self._on_connect:
                self._on_connect()

            # Start heartbeat and receive tasks
            self._heartbeat_task = asyncio.create_task(self._heartbeat_loop())
            self._receive_task = asyncio.create_task(self._receive_loop())
        except Exception as e:
            self.connected = False
            if self._on_error:
                self._on_error(e)
            await self._attempt_reconnect()

    def disconnect(self) -> None:
        """Disconnect from the WebSocket server."""
        self.connected = False

        # Cancel all tasks
        if self._heartbeat_task:
            self._heartbeat_task.cancel()
        if self._receive_task:
            self._receive_task.cancel()
        if self._reconnect_task:
            self._reconnect_task.cancel()

        # Close socket
        if self.socket:
            asyncio.create_task(self.socket.close())

    async def subscribe_prices(
        self, market_ids: list[str], interval_ms: int = 1000
    ) -> None:
        """
        Subscribe to price updates for specified markets.

        Args:
            market_ids: Market IDs to subscribe to
            interval_ms: Update interval in milliseconds
        """
        self._request_id += 1
        request = JsonRpcRequest(
            id=self._request_id,
            method="subscribe_prices",
            params={
                "channel": "prices",
                "market_ids": market_ids,
                "interval_ms": interval_ms,
            },
        )
        await self._send_request(request)

    async def subscribe_orderbook(
        self,
        market_ids: list[str],
        depth: int = 10,
        interval_ms: int = 2000,
    ) -> None:
        """
        Subscribe to orderbook updates for specified markets.

        Args:
            market_ids: Market IDs to subscribe to
            depth: Orderbook depth
            interval_ms: Update interval in milliseconds
        """
        self._request_id += 1
        request = JsonRpcRequest(
            id=self._request_id,
            method="subscribe_orderbook",
            params={
                "channel": "orderbook",
                "market_ids": market_ids,
                "depth": depth,
                "interval_ms": interval_ms,
            },
        )
        await self._send_request(request)

    async def unsubscribe(
        self, channel: str, market_ids: list[str]
    ) -> None:
        """
        Unsubscribe from a channel.

        Args:
            channel: Channel name ('prices' or 'orderbook')
            market_ids: Market IDs to unsubscribe from
        """
        self._request_id += 1
        request = JsonRpcRequest(
            id=self._request_id,
            method="unsubscribe",
            params={
                "channel": channel,
                "market_ids": market_ids,
            },
        )
        await self._send_request(request)

    async def get_market(self, market_id: str) -> Dict[str, Any]:
        """Get a specific market (one-off request)."""
        self._request_id += 1
        request = JsonRpcRequest(
            id=self._request_id,
            method="get_market",
            params={"market_id": market_id},
        )
        return await self._send_rpc_request(request)

    def is_connected(self) -> bool:
        """Check if WebSocket is connected."""
        return (
            self.connected
            and self.socket is not None
            and not self.socket.closed
        )

    # ─── Private Methods ────────────────────────────────────

    async def _send_request(self, request: JsonRpcRequest) -> None:
        """Send a JSON-RPC request."""
        if not self.is_connected():
            raise RuntimeError("WebSocket not connected")

        message = json.dumps(request.model_dump(exclude_none=True))
        await self.socket.send(message)

    async def _send_rpc_request(
        self, request: JsonRpcRequest
    ) -> Dict[str, Any]:
        """Send a JSON-RPC request and wait for response."""
        if not self.is_connected():
            raise RuntimeError("WebSocket not connected")

        message = json.dumps(request.model_dump(exclude_none=True))
        await self.socket.send(message)

        # Wait for response with timeout
        try:
            async for msg in self.socket:
                response = json.loads(msg)
                if response.get("id") == request.id:
                    rpc_response = JsonRpcResponse.model_validate(response)
                    if rpc_response.error:
                        raise RuntimeError(rpc_response.error["message"])
                    return rpc_response.result
        except asyncio.TimeoutError:
            raise RuntimeError(f"Request {request.id} timed out")

    async def _heartbeat_loop(self) -> None:
        """Send periodic heartbeat pings."""
        while self.is_connected():
            try:
                await asyncio.sleep(self.heartbeat_interval)
                if self.is_connected():
                    self._request_id += 1
                    request = JsonRpcRequest(
                        id=self._request_id, method="ping", params={}
                    )
                    await self._send_request(request)
            except Exception as e:
                if self._on_error:
                    self._on_error(e)

    async def _receive_loop(self) -> None:
        """Receive and handle messages from the WebSocket."""
        try:
            async for msg in self.socket:
                try:
                    data = json.loads(msg)

                    # Check if it's a fan-out message or JSON-RPC response
                    if "channel" in data:
                        if data["channel"] == "prices":
                            update = PriceUpdate.model_validate(data)
                            if self._on_price:
                                self._on_price(update)
                        elif data["channel"] == "orderbook":
                            update = OrderbookUpdate.model_validate(data)
                            if self._on_orderbook:
                                self._on_orderbook(update)
                except Exception as e:
                    if self._on_error:
                        self._on_error(e)
        except asyncio.CancelledError:
            pass
        except Exception as e:
            if self._on_error:
                self._on_error(e)
            await self._attempt_reconnect()

    async def _attempt_reconnect(self) -> None:
        """Attempt to reconnect with exponential backoff."""
        if not self.reconnect_enabled:
            return
        if self.reconnect_attempts >= self.max_reconnect_attempts:
            if self._on_error:
                self._on_error(
                    RuntimeError(
                        f"Failed to reconnect after {self.reconnect_attempts} attempts"
                    )
                )
            return

        delay = min(
            self.initial_reconnect_delay
            * (
                self.reconnect_backoff_multiplier
                ** self.reconnect_attempts
            ),
            self.max_reconnect_delay,
        )

        self.reconnect_attempts += 1
        await asyncio.sleep(delay)

        try:
            await self.connect()
        except Exception as e:
            if self._on_error:
                self._on_error(e)
            await self._attempt_reconnect()
